### PR TITLE
Reject methods called on unconstrained type_parameters

### PR DESCRIPTION
**Pull request by jez**
### Motivation

This only causes 96 errors on Stripe's codebase, and they all look like legitimate errors.

I've included an autocorrect to wrap the receiver in `T.unsafe` (pass the `--suggest-unsafe` flag at the command line in addition to the `--autocorrect` flag to enable the `T.unsafe` autocorrect).

I think we should move forward with this, as it takes a step towards making generics better at almost no cost.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_L9deCPYRkhG8V5
→ https://go/builds/bui_L9deLg7u2PHsef

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_L9eLGHwTKdLcAv
→ https://go/builds/bui_L9eLMyf8eSKxb7

File: core/types/calls.cc [hunk 1]
```diff
@@ -137,6 +137,53 @@ DispatchResult TupleType::dispatchCall(const GlobalState &gs, const DispatchArgs
...
+                }
+            }
+        }
+        result.main.errors.emplace_back(e.build());
+        return result;
+    } else {
+        // TODO: https://github.com/sorbet/sorbet/issues/1731
+        auto untypedUntracked = Types::untypedUntracked();
+        return untypedUntracked.dispatchCall(gs, args.withThisRef(untypedUntracked));
+    }
```
Comment (jez) [inline core/types/calls.cc:184]: @elliottt question for you: Do you have an intuition for why `SelfTypeParam` models both `type_member` symbols and `type_parameter` symbols?

There's some code over here in `isSubTypeUnderConstraintSingle` that seems to do something similar (different logic depending on type argument vs type member):

https://github.com/sorbet/sorbet/blob/96c4d8e18db19a677b96e3edc41bcd9fa1365781/core/types/subtyping.cc#L1096-L1129

Comment (elliottt) [reply core/types/calls.cc:186]: I've always thought of `SelfTypeParam` as a skolem variable, and using it to represent both a `type_member` and `type_parameter` is probably just for convenience.

In this case in subtyping.cc, we're using the fact that `LambdaParam` has bounds information present to discover cases where we know they must be compatible. In all other cases they're considered to be incompatible. [Here's an example](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20Parent%3B%20end%0Aclass%20Child%20%3C%20Parent%3B%20end%0A%0Aclass%20WithBounds%0A%20%20extend%20T%3A%3AGeneric%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20A%20%3D%20type_member%28upper%3A%20Child%29%0A%20%20B%20%3D%20type_member%28lower%3A%20Parent%29%0A%0A%20%20sig%20%7Breturns%28A%29%7D%0A%20%20def%20make_A%0A%20%20%20%20T.unsafe%28nil%29%0A%20%20end%0A%0A%20%20sig%20%7Bparams%28b%3A%20B%29.void%7D%0A%20%20def%20uses_B%28b%29%0A%20%20end%0A%0A%20%20def%20combine%0A%20%20%20%20a%20%3D%20self.make_A%0A%20%20%20%20self.uses_B%28a%29%0A%20%20end%0A%0Aend%0A%0Aclass%20WithoutBounds%0A%20%20extend%20T%3A%3AGeneric%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20A%20%3D%20type_member%28%29%0A%20%20B%20%3D%20type_member%28%29%0A%0A%20%20sig%20%7Breturns%28A%29%7D%0A%20%20def%20make_A%0A%20%20%20%20T.unsafe%28nil%29%0A%20%20end%0A%0A%20%20sig%20%7Bparams%28b%3A%20B%29.void%7D%0A%20%20def%20uses_B%28b%29%0A%20%20end%0A%0A%20%20def%20combine%0A%20%20%20%20a%20%3D%20self.make_A%0A%20%20%20%20self.uses_B%28a%29%0A%20%20end%0A%0Aend) where the bounds information allows interactions of `type_member` values in a context where they're skolemized.

**Review by aprocter — APPROVED** [review]

**Pull request merged**